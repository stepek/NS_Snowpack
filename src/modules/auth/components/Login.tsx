import React, {useCallback, useState} from "react"
import {useDispatch, useSelector} from "react-redux"

import {Button, TextField, Typography} from "../../../components"
import {loginAction} from "../authDuck/actions"
import {getErrorMessage} from "../authDuck/selectors"

function parseLoginError(error: string): string {
  if (error === "Unauthorized") {
    return "The username or password is incorrect."
  }
  return error
}

export function Login() {
  const dispatch = useDispatch()
  const [username, setUsername] = useState<string>("")
  const [password, setPassword] = useState<string>("")
  const errorMessage = useSelector(getErrorMessage)

  const handleLoginClick = useCallback(() => {
    dispatch(loginAction(username, password))
  }, [dispatch, username, password])

  return (
    <>
      {errorMessage !== null && (
        <Typography>{parseLoginError(errorMessage)}</Typography>
      )}
      <TextField
        placeholder={"User Name"}
        value={username}
        onChange={setUsername}
      />
      <TextField
        placeholder={"Password"}
        value={password}
        onChange={setPassword}
      />
      <Button label={"Login"} onClick={handleLoginClick} />
    </>
  )
}

export default Login
