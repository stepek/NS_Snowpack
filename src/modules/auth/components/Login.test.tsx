import React from "react"
import {useSelector, useDispatch} from "react-redux"
import renderer from "react-test-renderer"

import Login from "./Login"

jest.mock("react-redux")

afterEach(() => {
  useSelector.mockClear()
  useDispatch.mockClear()
})

test("<Login> render correctly", () => {
  useSelector.mockImplementationOnce(() => null)
  useDispatch.mockImplementationOnce(() => undefined)

  expect(renderer.create(<Login />).toJSON()).toMatchSnapshot()
})

test("<Login> render correctly with error", () => {
  useSelector.mockImplementationOnce(() => "Unauthorized")
  useDispatch.mockImplementationOnce(() => undefined)

  expect(renderer.create(<Login />).toJSON()).toMatchSnapshot()
})

test("<Login> render correctly with unknown error", () => {
  useSelector.mockImplementationOnce(() => "unknown error")
  useDispatch.mockImplementationOnce(() => undefined)

  expect(renderer.create(<Login />).toJSON()).toMatchSnapshot()
})
