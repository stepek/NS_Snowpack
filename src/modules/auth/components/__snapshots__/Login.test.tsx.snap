// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login> render correctly 1`] = `
Array [
  <input
    onChange={[Function]}
    placeholder="User Name"
    value=""
  />,
  <input
    onChange={[Function]}
    placeholder="Password"
    value=""
  />,
  <button
    onClick={[Function]}
  >
    Login
  </button>,
]
`;

exports[`<Login> render correctly with error 1`] = `
Array [
  <span>
    The username or password is incorrect.
  </span>,
  <input
    onChange={[Function]}
    placeholder="User Name"
    value=""
  />,
  <input
    onChange={[Function]}
    placeholder="Password"
    value=""
  />,
  <button
    onClick={[Function]}
  >
    Login
  </button>,
]
`;

exports[`<Login> render correctly with unknown error 1`] = `
Array [
  <span>
    unknown error
  </span>,
  <input
    onChange={[Function]}
    placeholder="User Name"
    value=""
  />,
  <input
    onChange={[Function]}
    placeholder="Password"
    value=""
  />,
  <button
    onClick={[Function]}
  >
    Login
  </button>,
]
`;
